pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +234,295 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+	// Controller will create corresponding interface and route on the host.
+	if !c.wireGuardInitialized && gateway != nil && gateway.Name == c.nodeConfig.Name {
+		klog.InfoS("Initializing WireGuard interface and rule")
+		var initErr error
+		defer func() {
+			if initErr != nil {
+				klog.InfoS("Failed to initialize WireGuard, deleting interface and route")
+				if deleteErr := c.deleteWireGuardInterfaceAndRule(); deleteErr != nil {
+					klog.ErrorS(deleteErr, "Failed to cleanup WireGuard stale resources")
+				}
+			}
+		}()
+
+		if initErr = c.initializeWireGuard(); initErr != nil {
+			return false, err
+		}
+		if initErr = c.initWireGuardRoute(); initErr != nil {
+			return false, initErr
+		}
+		c.wireGuardInitialized = true
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// deleteWireGuardInterfaceAndRule deletes the WireGuard interfaces and IP rule on the host.
+// The IP route will also be deleted automatically when the interfaces deleted.
+// deleteWireGuardInterfaceAndRule supports reentrance, if the users have deleted the IP rule or
+// interface manually, it won't return an error when deleted those resources 2nd time.
+func (c *MCDefaultRouteController) deleteWireGuardInterfaceAndRule() error {
+	if err := c.routeClient.DeleteIPRule(multiclusterWireGuardFwmark, multiclusterWireGuardRouteTable); err != nil &&
+		!strings.Contains(err.Error(), "no such file or directory") {
+		// If the error message contains "no such file or directory", it means that the IP rule does not exist,
+		// so just return nil.
+		return err
+	}
+	if err := c.wireGuardClient.Cleanup(); err != nil {
+		return err
+	}
+	c.wireGuardClient = nil
+	return nil
+}
+
+// initWireGuardRoute initializes the WireGuard IP rule and route on the host.
+// If the Rule or route already exists, it will do nothing.
+func (c *MCDefaultRouteController) initWireGuardRoute() error {
+	if err := c.routeClient.AddIPRule(multiclusterWireGuardFwmark, multiclusterWireGuardRouteTable); err != nil {
+		return err
+	}
+	if err := c.routeClient.AddRouteToTable(multiclusterWireGuardInterface, net.IPv4zero, multiclusterWireGuardRouteTable); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MCDefaultRouteController) updateWireGuardPeers(ciImp string) error {
+	c.ciImportMutex.Lock()
+	defer c.ciImportMutex.Unlock()
+	if !c.wireGuardInitialized {
+		return fmt.Errorf("failed to update WireGuard peers, WireGuard has not been initialized")
+	}
+
+	if ciImp != "" {
+		ciImport, err := c.ciImportLister.ClusterInfoImports(c.namespace).Get(ciImp)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			klog.InfoS("The ClusterInfoImport has been deleted, removing WireGuard peer", "ClusterInfoImport", ciImp)
+			return c.wireGuardClient.DeletePeer(ciImp)
+		}
+		return c.updateSinglePeer(ciImport)
+	}
+
+	// The ClusterInfoImport's name is empty. It means that a new Gateway is created, controller should add all existing ClusterInfoImport
```

Comment:
I believe it is much easier to have a different func for this case. Nothing shared, and locking context, etc. are different for the two cases.